src/storage/src/hummock/store/version.rs

```diff
@@ -546,9 +620,23 @@ impl HummockVersionReader {
         table_key: TableKey<Bytes>,
         epoch: u64,
         read_options: ReadOptions,
-        read_version_tuple: (Vec<ImmutableMemtable>, Vec<SstableInfo>, CommittedVersion),
+        read_version_tuple: ReadVersionTuple,
     ) -> StorageResult<Option<Bytes>> {
-        let (imms, uncommitted_ssts, committed_version) = read_version_tuple;
+        let (imms, uncommitted_ssts, committed_version, watermark) = read_version_tuple;
+        let key_vnode = table_key.vnode_part();
+        if let Some(read_watermark) = watermark {
+            for (vnode, watermark) in read_watermark.vnode_watermarks {
```

Comment:
`vnode_watermarks` is a BTreeMap. We can do binary search instead of linear search.